README.md
This package has moved to https://github.com/dart-lang/shelf/tree/master/pkgs/shelf_proxy


CHANGELOG.md
## 1.0.0

- Require Dart `2.12`.
- Enable null safety.

## 0.1.0+7

* Added example.
* Fixed links in README.

## 0.1.0+6

* Support the latest version of `package:http`.

## 0.1.0+5

* Support Dart 2.

## 0.1.0+4

* Internal changes only.

## 0.1.0+3

* Support version `0.7.0` of `shelf`.

## 0.1.0+2

* Support version `0.6.0` of `shelf`.

## 0.1.0+1

* Added `drone.io` badge to `README.md`.

## 0.1.0

* `createProxyHandler` (new deprecated) is replaced with `proxyHandler`.

* Updated to be compatible with RFC 2616 Proxy specification.

## 0.0.2

* Updated `README.md` and doc comments on `createProxyHandler`.

* Added an example.

## 0.0.1

* First release.


test/shelf_proxy_test.dart
// Copyright (c) 2014, the Dart project authors.  Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.

import 'dart:async';

import 'package:http/http.dart' as http;
import 'package:http/testing.dart';
import 'package:shelf/shelf.dart' as shelf;
import 'package:shelf/shelf_io.dart' as shelf_io;
import 'package:shelf_proxy/shelf_proxy.dart';
import 'package:test/test.dart';

/// The URI of the server the current proxy server is proxying to.
late Uri targetUri;

/// The URI of the current proxy server.
late Uri proxyUri;

void main() {
  group('forwarding', () {
    test('forwards request method', () async {
      await createProxy((request) {
        expect(request.method, equals('DELETE'));
        return shelf.Response.ok(':)');
      });

      await http.delete(proxyUri);
    });

    test('forwards request headers', () async {
      await createProxy((request) {
        expect(request.headers, containsPair('foo', 'bar'));
        expect(request.headers, containsPair('accept', '*/*'));
        return shelf.Response.ok(':)');
      });

      await get(headers: {'foo': 'bar', 'accept': '*/*'});
    });

    test('forwards request body', () async {
      await createProxy((request) {
        expect(request.readAsString(), completion(equals('hello, server')));
        return shelf.Response.ok(':)');
      });

      await http.post(proxyUri, body: 'hello, server');
    });

    test('forwards response status', () async {
      await createProxy((request) {
        return shelf.Response(567);
      });

      var response = await get();
      expect(response.statusCode, equals(567));
    });

    test('forwards response headers', () async {
      await createProxy((request) {
        return shelf.Response.ok(':)',
            headers: {'foo': 'bar', 'accept': '*/*'});
      });

      var response = await get();

      expect(response.headers, containsPair('foo', 'bar'));
      expect(response.headers, containsPair('accept', '*/*'));
    });

    test('forwards response body', () async {
      await createProxy((request) {
        return shelf.Response.ok('hello, client');
      });

      expect(await http.read(proxyUri), equals('hello, client'));
    });

    test('adjusts the Host header for the target server', () async {
      await createProxy((request) {
        expect(request.headers, containsPair('host', targetUri.authority));
        return shelf.Response.ok(':)');
      });

      await get();
    });
  });

  group('via', () {
    test('adds a Via header to the request', () async {
      await createProxy((request) {
        expect(request.headers, containsPair('via', '1.1 shelf_proxy'));
        return shelf.Response.ok(':)');
      });

      await get();
    });

    test("adds to a request's existing Via header", () async {
      await createProxy((request) {
        expect(request.headers,
            containsPair('via', '1.0 something, 1.1 shelf_proxy'));
        return shelf.Response.ok(':)');
      });

      await get(headers: {'via': '1.0 something'});
    });

    test('adds a Via header to the response', () async {
      await createProxy((request) => shelf.Response.ok(':)'));

      var response = await get();
      expect(response.headers, containsPair('via', '1.1 shelf_proxy'));
    });

    test("adds to a response's existing Via header", () async {
      await createProxy((request) {
        return shelf.Response.ok(':)', headers: {'via': '1.0 something'});
      });

      var response = await get();
      expect(response.headers,
          containsPair('via', '1.0 something, 1.1 shelf_proxy'));
    });
  });

  group('redirects', () {
    test("doesn't modify a Location for a foreign server", () async {
      await createProxy((request) {
        return shelf.Response.found('http://dartlang.org');
      });

      var response = await get();
      expect(response.headers, containsPair('location', 'http://dartlang.org'));
    });

    test('relativizes a reachable root-relative Location', () async {
      await createProxy((request) {
        return shelf.Response.found('/foo/bar');
      }, targetPath: '/foo');

      var response = await get();
      expect(response.headers, containsPair('location', '/bar'));
    });

    test('absolutizes an unreachable root-relative Location', () async {
      await createProxy((request) {
        return shelf.Response.found('/baz');
      }, targetPath: '/foo');

      var response = await get();
      expect(response.headers,
          containsPair('location', targetUri.resolve('/baz').toString()));
    });
  });

  test('removes a transfer-encoding header', () async {
    var handler = mockHandler((request) {
      return http.Response('', 200, headers: {'transfer-encoding': 'chunked'});
    });

    var response =
        await handler(shelf.Request('GET', Uri.parse('http://localhost/')));

    expect(response.headers, isNot(contains('transfer-encoding')));
  });

  test('removes content-length and content-encoding for a gzipped response',
      () async {
    var handler = mockHandler((request) {
      return http.Response('', 200,
          headers: {'content-encoding': 'gzip', 'content-length': '1234'});
    });

    var response =
        await handler(shelf.Request('GET', Uri.parse('http://localhost/')));

    expect(response.headers, isNot(contains('content-encoding')));
    expect(response.headers, isNot(contains('content-length')));
    expect(response.headers,
        containsPair('warning', '214 shelf_proxy "GZIP decoded"'));
  });
}

/// Creates a proxy server proxying to a server running [handler].
///
/// [targetPath] is the root-relative path on the target server to proxy to. It
/// defaults to `/`.
Future createProxy(shelf.Handler handler, {String? targetPath}) async {
  handler = expectAsync1(handler, reason: 'target server handler');
  var targetServer = await shelf_io.serve(handler, 'localhost', 0);
  targetUri = Uri.parse('http://localhost:${targetServer.port}');
  if (targetPath != null) targetUri = targetUri.resolve(targetPath);
  var proxyServerHandler =
      expectAsync1(proxyHandler(targetUri), reason: 'proxy server handler');

  var proxyServer = await shelf_io.serve(proxyServerHandler, 'localhost', 0);
  proxyUri = Uri.parse('http://localhost:${proxyServer.port}');

  addTearDown(() {
    proxyServer.close(force: true);
    targetServer.close(force: true);
  });
}

/// Creates a [shelf.Handler] that's backed by a [MockClient] running
/// [callback].
shelf.Handler mockHandler(
    FutureOr<http.Response> Function(http.Request) callback) {
  var client = MockClient((request) async => await callback(request));
  return proxyHandler('http://dartlang.org', client: client);
}

/// Schedules a GET request with [headers] to the proxy server.
Future<http.Response> get({Map<String, String>? headers}) {
  var uri = proxyUri;
  var request = http.Request('GET', uri);
  if (headers != null) request.headers.addAll(headers);
  request.followRedirects = false;
  return request.send().then(http.Response.fromStream);
}


lib/shelf_proxy.dart
// Copyright (c) 2014, the Dart project authors.  Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.

import 'package:http/http.dart' as http;
import 'package:path/path.dart' as p;
import 'package:pedantic/pedantic.dart';
import 'package:shelf/shelf.dart';

/// A handler that proxies requests to [url].
///
/// To generate the proxy request, this concatenates [url] and [Request.url].
/// This means that if the handler mounted under `/documentation` and [url] is
/// `http://example.com/docs`, a request to `/documentation/tutorials`
/// will be proxied to `http://example.com/docs/tutorials`.
///
/// [url] must be a [String] or [Uri].
///
/// [client] is used internally to make HTTP requests. It defaults to a
/// `dart:io`-based client.
///
/// [proxyName] is used in headers to identify this proxy. It should be a valid
/// HTTP token or a hostname. It defaults to `shelf_proxy`.
Handler proxyHandler(url, {http.Client? client, String? proxyName}) {
  Uri uri;
  if (url is String) {
    uri = Uri.parse(url);
  } else if (url is Uri) {
    uri = url;
  } else {
    throw ArgumentError.value(url, 'url', 'url must be a String or Uri.');
  }
  final theClient = client ?? http.Client();
  proxyName ??= 'shelf_proxy';

  return (serverRequest) async {
    // TODO(nweiz): Support WebSocket requests.

    // TODO(nweiz): Handle TRACE requests correctly. See
    // http://www.w3.org/Protocols/rfc2616/rfc2616-sec9.html#sec9.8
    var requestUrl = uri.resolve(serverRequest.url.toString());
    var clientRequest = http.StreamedRequest(serverRequest.method, requestUrl);
    clientRequest.followRedirects = false;
    clientRequest.headers.addAll(serverRequest.headers);
    clientRequest.headers['Host'] = uri.authority;

    // Add a Via header. See
    // http://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.45
    _addHeader(clientRequest.headers, 'via',
        '${serverRequest.protocolVersion} $proxyName');

    unawaited(serverRequest
        .read()
        .forEach(clientRequest.sink.add)
        .catchError(clientRequest.sink.addError)
        .whenComplete(clientRequest.sink.close));
    var clientResponse = await theClient.send(clientRequest);
    // Add a Via header. See
    // http://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.45
    _addHeader(clientResponse.headers, 'via', '1.1 $proxyName');

    // Remove the transfer-encoding since the body has already been decoded by
    // [client].
    clientResponse.headers.remove('transfer-encoding');

    // If the original response was gzipped, it will be decoded by [client]
    // and we'll have no way of knowing its actual content-length.
    if (clientResponse.headers['content-encoding'] == 'gzip') {
      clientResponse.headers.remove('content-encoding');
      clientResponse.headers.remove('content-length');

      // Add a Warning header. See
      // http://www.w3.org/Protocols/rfc2616/rfc2616-sec13.html#sec13.5.2
      _addHeader(
          clientResponse.headers, 'warning', '214 $proxyName "GZIP decoded"');
    }

    // Make sure the Location header is pointing to the proxy server rather
    // than the destination server, if possible.
    if (clientResponse.isRedirect &&
        clientResponse.headers.containsKey('location')) {
      var location =
          requestUrl.resolve(clientResponse.headers['location']!).toString();
      if (p.url.isWithin(uri.toString(), location)) {
        clientResponse.headers['location'] =
            '/' + p.url.relative(location, from: uri.toString());
      } else {
        clientResponse.headers['location'] = location;
      }
    }

    return Response(clientResponse.statusCode,
        body: clientResponse.stream, headers: clientResponse.headers);
  };
}

/// Use [proxyHandler] instead.
@deprecated
Handler createProxyHandler(Uri rootUri) => proxyHandler(rootUri);

// TODO(nweiz): use built-in methods for this when http and shelf support them.
/// Add a header with [name] and [value] to [headers], handling existing headers
/// gracefully.
void _addHeader(Map<String, String> headers, String name, String value) {
  final existing = headers[name];
  headers[name] = existing == null ? value : '$existing, $value';
}
